algebra/group/default.lean

```diff
@@ -484,23 +484,31 @@ section add_comm_group
 
 end add_comm_group
 
-variables {β : Type*} [group α] [group β] {a b : α}
+variables  [group α] {β : Type*} [group β]
 
 /-- Predicate for group homomorphism. -/
-def is_group_hom (f : α → β) : Prop :=
-∀ a b : α, f (a * b) = f a * f b
+class is_group_hom (f : α → β) :=
+  (hom_mul : ∀ a b : α, f (a * b) = f a * f b)
 
 namespace is_group_hom
-variables {f : α → β} (H : is_group_hom f)
-include H
+variables (f : α → β) [w : is_group_hom f]
+include w
 
-theorem mul : ∀ a b : α, f (a * b) = f a * f b := H
+@[simp] theorem mul : ∀ a b : α, f (a * b) = f a * f b := w.hom_mul
```

Comment:
These can't be simp lemmas, they don't work since `f` is a variable.

Reply:
I wish I understood simp better. I thought that simp was fine with "(precondition) -> X = Y", isn't that what is going on here? forall f, H (saying f is a group hom), a and b, blah?

Reply:
The reason why `mul` won't work (well[1]) as a simp lemma is not because there is a precondition, but because the function symbol `f` is not a constant (it is a variable here).

The simplifier indexes the available lemmas by the head function symbol of the left-hand side, for example `foo 0 = 1` would be indexed by the name `foo`.  If the simplifier then sees a subterm of the form `foo (a*b)`, it will only look for simp lemmas for the name `foo` and will not find the `mul` lemma.

[1] I believe that you can use `mul` as a simp lemma if the homomorphism that you want to rewrite is a local constant, but I haven't tested it.

Reply:
Thanks!